src/protocol/loli_nrf24l01.c

```diff
@@ -0,0 +1,420 @@
+
+#include "iface_nrf24l01.h"
+
+#define LOLI_PACKET_SIZE    11
+#define LOLI_BIND_CHANNEL   33
+#define LOLI_NUM_CHANNELS   5
+
+static u8 packet[LOLI_PACKET_SIZE];
+static u8 phase;
+static u8 tx_power;
+static u8 hopping_frequency[LOLI_NUM_CHANNELS];
+static u8 rx_tx_addr[5];
+static u8 hopping_frequency_no;
+static u8 count;
+static u16 rx_config;
+static const u8 bind_address[5] = {'L', 'O', 'V', 'E', '!'};
+
+enum{
+    BIND1,
+    BIND2,
+    BIND3,
+    DATA1,
+    DATA2,
+    SET_RX_CONFIG
+};
+
+static const char * const loli_opts[] = {
+    _tr_noop("Chan 1"), "Servo", "SW", "PWM", "PPM", NULL,
+    _tr_noop("Chan 2"), "Servo", "SW", "PWM", NULL,
+    _tr_noop("Chan 3"), "Servo", "SW", NULL,
+    _tr_noop("Chan 4"), "Servo", "SW", NULL,
+    _tr_noop("Chan 5"), "Servo", "SW", "SBUS", NULL,
+    _tr_noop("Chan 6"), "Servo", "SW", NULL,
+    _tr_noop("Chan 7"), "Servo", "SW", "PWM", NULL,
+    _tr_noop("Chan 8"), "Servo", "SW", NULL,
+    NULL
+};
+
+#define OPT_SERVO   0
+#define OPT_SW      1
+#define OPT_PWM     2
+#define OPT_SBUS    2
+#define OPT_PPM     3
+
+enum {
+    PROTOOPTS_CH1 = 0,
+    PROTOOPTS_CH2,
+    PROTOOPTS_CH3,
+    PROTOOPTS_CH4,
+    PROTOOPTS_CH5,
+    PROTOOPTS_CH6,
+    PROTOOPTS_CH7,
+    PROTOOPTS_CH8,
+    LAST_PROTO_OPT,
+};
+ctassert(LAST_PROTO_OPT <= NUM_PROTO_OPTS, too_many_protocol_opts);
+
+// flags going to packet[1] for packet type 0xa2 (Rx config)
+#define FLAG_PWM7   0x02
+#define FLAG_PWM2   0x04
+#define FLAG_PWM1   0x08
+#define FLAG_SBUS   0x40
+#define FLAG_PPM    0x80
+
+// flags going to packet[2] for packet type 0xa2 (Rx config)
+#define FLAG_SW8    0x01
+#define FLAG_SW7    0x02
+#define FLAG_SW6    0x04
+#define FLAG_SW5    0x08
+#define FLAG_SW4    0x10
+#define FLAG_SW3    0x20
+#define FLAG_SW2    0x40
+#define FLAG_SW1    0x80
+
+// For code readability
+enum {
+    CHANNEL1 = 0,
+    CHANNEL2,
+    CHANNEL3,
+    CHANNEL4,
+    CHANNEL5,
+    CHANNEL6,
+    CHANNEL7,
+    CHANNEL8,
+};
+
+// Bit vector from bit position
+#define BV(bit) (1 << bit)
+
+static void init_RF()
+{
+    NRF24L01_Initialize();
+    NRF24L01_FlushTx();
+    NRF24L01_FlushRx();
+    NRF24L01_WriteReg(NRF24L01_07_STATUS, 0x70);     // Clear data ready, data sent, and retransmit
+    NRF24L01_WriteReg(NRF24L01_01_EN_AA, 0x00);      // No Auto Acknowldgement on all data pipes
+    NRF24L01_WriteReg(NRF24L01_02_EN_RXADDR, 0x01);  // enable rx data pipe 0
+    NRF24L01_WriteReg(NRF24L01_04_SETUP_RETR, 0x00);    // No retransmit
+    NRF24L01_WriteReg(NRF24L01_05_RF_CH, 66);   // start frequency
+    NRF24L01_WriteReg(NRF24L01_11_RX_PW_P0, LOLI_PACKET_SIZE);  // RX FIFO size
+    NRF24L01_SetBitrate(NRF24L01_BR_250K);
+    tx_power = Model.tx_power;
+    NRF24L01_SetPower(tx_power);
+    NRF24L01_SetTxRxMode(TX_EN);
+    NRF24L01_WriteReg(NRF24L01_00_CONFIG, 0x0a);  // 8bit CRC, TX
+}
+
+#define CHAN_RANGE (CHAN_MAX_VALUE - CHAN_MIN_VALUE)
+static u16 scale_channel(u8 ch, u16 destMin, u16 destMax)
```

Comment:
better pass chanval into this function so that this function is a purely matchmatical function and can be easier to test.

Reply:
Agree, I'll also pass CHAN_MAX_VALUE & CHAN_MAX_VALUE so the same function can be used to scale failsafe settings as well.